diff --git a/tmk_core/protocol.mk b/tmk_core/protocol.mk
index d159dacc55..dc87e88ef4 100644
--- a/tmk_core/protocol.mk
+++ b/tmk_core/protocol.mk
@@ -39,10 +39,6 @@ endif
 
 ifeq ($(strip $(CONSOLE_ENABLE)), yes)
     OPT_DEFS += -DCONSOLE_ENABLE
-else
-    # TODO: decouple this so other print backends can exist
-    OPT_DEFS += -DNO_PRINT
-    OPT_DEFS += -DNO_DEBUG
 endif
 
 ifeq ($(strip $(NKRO_ENABLE)), yes)
