diff --git a/quantum/painter/qp_draw_core.c b/quantum/painter/qp_draw_core.c
index 2952825200..3fe61b0f6d 100644
--- a/quantum/painter/qp_draw_core.c
+++ b/quantum/painter/qp_draw_core.c
@@ -137,14 +137,14 @@ bool qp_internal_load_qgf_palette(qp_stream_t *stream, uint8_t bpp, uint8_t tran
 
         qp_internal_global_pixel_lookup_table[i].transparency = i == transparency_index;
 
-        qp_dprintf(
-            "qp_internal_load_qgf_palette: %3d of %d -- H: %3d, S: %3d, V: %3d -- transparent: %s\n",
+        dprintf(
+            "%3d/%d - H:%3d, S:%3d, V:%3d|%s\n",
             (int)(i + 1),
             (int)palette_entries,
             (int)qp_internal_global_pixel_lookup_table[i].hsv888.h,
             (int)qp_internal_global_pixel_lookup_table[i].hsv888.s,
             (int)qp_internal_global_pixel_lookup_table[i].hsv888.v,
-            qp_internal_global_pixel_lookup_table[i].transparency ? "true" : "false"
+            qp_internal_global_pixel_lookup_table[i].transparency ? "yes" : "no"
         );
     }
 
